src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -0,0 +1,581 @@
+                _devMinor = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(buffer.Slice(FieldLocations.DevMinor, FieldLengths.DevMinor));
+            }
+        }
+
+        // Reads attributes specific to the GNU format.
+        // Throws if any conversion fails.
+        private void ReadGnuAttributes(Span<byte> buffer)
+        {
+            // Convert byte arrays
+            int aTime = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(buffer.Slice(FieldLocations.ATime, FieldLengths.ATime));
+            _aTime = TarHelpers.GetDateTimeFromSecondsSinceEpoch(aTime);
+
+            int cTime = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(buffer.Slice(FieldLocations.CTime, FieldLengths.CTime));
+            _cTime = TarHelpers.GetDateTimeFromSecondsSinceEpoch(cTime);
+
+            // TODO: Read the bytes of the currently unsupported GNU fields, in case user wants to write this entry into another GNU archive, they need to be preserved.
+        }
+
+        // Reads the ustar prefix attribute.
+        // Throws if a conversion to an expected data type fails.
+        private void ReadUstarAttributes(Span<byte> buffer)
+        {
+            _prefix = TarHelpers.GetTrimmedUtf8String(buffer.Slice(FieldLocations.Prefix, FieldLengths.Prefix));
+
+            // In ustar, Prefix is used to store the *leading* path segments of
+            // Name, if the full path did not fit in the Name byte array.
+            if (!string.IsNullOrEmpty(_prefix))
+            {
+                // Prefix never has a leading separator, so we add it
+                // it should always  be a forward slash for compatibility
+                _name = string.Format(UstarPrefixFormat, _prefix, _name);
+            }
+        }
+
+        // Collects the extended attributes found in the data section of a PAX entry of type 'x' or 'g'.
+        // Throws if end of stream is reached or if an attribute is malformed.
+        private void ReadExtendedAttributesBlock(Stream archiveStream)
+        {
+            Debug.Assert(_typeFlag is TarEntryType.ExtendedAttributes or TarEntryType.GlobalExtendedAttributes);
+
+            // Regardless of the size, this entry should always have a valid dictionary object
+            _extendedAttributes ??= new Dictionary<string, string>();
+
+            if (_size == 0)
+            {
+                return;
+            }
+
+            // It is not expected that the extended attributes data section will be longer than int.MaxValue, considering
+            // 4096 is a common max path length, and also the size field is 12 bytes long, which is under int.MaxValue.
+            if (_size > int.MaxValue)
+            {
+                throw new InvalidOperationException(string.Format(SR.TarSizeFieldTooLargeForExtendedAttribute, _typeFlag.ToString()));
+            }
+
+            byte[] buffer = new byte[(int)_size];
+            if (archiveStream.Read(buffer.AsSpan()) != _size)
```

Comment:
There's no guarantee that a Read on a Stream will return the exact amount asked for.  Is this stream special in some way that you know it makes that guarantee?

Reply:
Good catch. I should adjust this to look more like what I do inside the `ReadOrThrow` method.